### PR TITLE
Add ballerina services and listeners API

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
$subject

## Automation tests
- Unit tests
  > Code coverage information
- Integration tests
  > Details about the test cases and coverage

## Security checks
- Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
- Ran FindSecurityBugs plugin and verified report? yes/no
- Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: ballerina/types.bal [hunk 1]
```diff
@@ -38,41 +39,65 @@ type AccessTokenResponse record {|
...
+
+public type RequestLimit record {
+    int maxUriLength;
+    int maxHeaderSize;
+    int maxEntityBodySize;
+};
+
+public type ListenerDetail record {
+    string package;
+    string httpVersion;
```
Comment (warunalakshitha) [inline ballerina/types.bal]: Shall we move this fields to native side based on the protocol. This type can be open

File: native/src/main/java/io/ballerina/lib/wso2/controlplane/ArtifactUtils.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    public static Object getDetailedArtifact(BString resourceType, BString name) {
+        String value = name.getValue();
+        if (resourceType.getValue().equals(SERVICES_RESOURCE)) {
+           Artifact artifact = getServiceArtifact(value);
+           if (artifact == null) {
+               return ErrorCreator.createError(StringUtils.fromString("No service found with the name: " + name));
+           }
+           return serviceArtifactHandler.getDetailedService(artifact, currentModule);
+
```
Comment (gabilang) [inline native/src/main/java/io/ballerina/lib/wso2/controlplane/ArtifactUtils.java]:
```suggestion
```

File: native/src/main/java/io/ballerina/lib/wso2/controlplane/ServiceArtifactHandler.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            }
+            BMap<BString, Object> service = ValueCreator.createMapValue();
+            service.put(StringUtils.fromString(NAME), StringUtils.fromString(SERVICE_NAMES_MAP.get(serviceObj)));
+            service.put(StringUtils.fromString(BASE_PATH), getAttachPointString(artifact));
+            artifactEntries.add(ValueCreator.createListInitialValueEntry(
+                    ValueCreator.createReadonlyRecordValue(currentModule, SERVICE_RECORD, service)));
+        }
+        return artifactEntries;
+    }
+    public BMap<BString, Object> getDetailedService(Artifact artifact, Module currentModule) {
```
Comment (gabilang) [inline native/src/main/java/io/ballerina/lib/wso2/controlplane/ServiceArtifactHandler.java]:
```suggestion
    }
    
    public BMap<BString, Object> getDetailedService(Artifact artifact, Module currentModule) {
```

File: native/src/main/java/io/ballerina/lib/wso2/controlplane/ArtifactUtils.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+ */
+public class ArtifactUtils {
+    private static final String SERVICE_PREFIX = "service_";
+    private static final String LISTENER_PREFIX = "listener_";
+    static List<Artifact> artifacts;
+    private static Module currentModule;
+
+    private static final ServiceArtifactHandler serviceArtifactHandler = new ServiceArtifactHandler();
+    private static final ListenerArtifactHandler listenerArtifactHandler = new ListenerArtifactHandler();
+
```
Comment (gabilang) [inline native/src/main/java/io/ballerina/lib/wso2/controlplane/ArtifactUtils.java]:
```suggestion
```

File: native/src/main/java/io/ballerina/lib/wso2/controlplane/ArtifactUtils.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    static final Map<Object, String> SERVICE_NAMES_MAP = new HashMap<>();
+    static final Map<Object, String> LISTENER_NAMES_MAP = new HashMap<>();
+
```
Comment (gabilang) [inline native/src/main/java/io/ballerina/lib/wso2/controlplane/ArtifactUtils.java]:
```suggestion
```

File: ballerina/Ballerina.toml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 [package]
 org = "ballerinax"
 name = "wso2.controlplane"
-version = "1.0.0"
+version = "0.0.0"
```
Comment (warunalakshitha) [inline ballerina/Ballerina.toml]: Shall we use 0.1.0

File: gradle.properties [hunk 7]
```diff
@@ -1,29 +1,30 @@
 org.gradle.caching=true
 group=org.ballerinalang
-version=1.0.0-SNAPSHOT
+version=0.0.0-SNAPSHOT
```
Comment (warunalakshitha) [inline gradle.properties]:
```suggestion
version=0.1.0-SNAPSHOT
```

File: ballerina/Ballerina.toml [hunk 8]
```diff
@@ -10,4 +10,4 @@ license = ["Apache-2.0"]
 graalvmCompatible = true
 
 [[platform.java17.dependency]]
-path = "../native/build/libs/wso2.controlplane-native-1.0.0-SNAPSHOT.jar"
+path = "../native/build/libs/wso2.controlplane-native-0.0.0-SNAPSHOT.jar"
```
Comment (warunalakshitha) [inline ballerina/Ballerina.toml]:
```suggestion
path = "../native/build/libs/wso2.controlplane-native-0.1.0-SNAPSHOT.jar"
```

File: ballerina/Dependencies.toml [hunk 9]
```diff
@@ -286,18 +286,33 @@ dependencies = [
...
+]
+modules = [
+	{org = "ballerina", packageName = "uuid", moduleName = "uuid"}
+]
+
 [[package]]
 org = "ballerinax"
 name = "wso2.controlplane"
-version = "1.0.0"
+version = "0.0.0"
```
Comment (warunalakshitha) [inline ballerina/Dependencies.toml]:
```suggestion
version = "0.1.0"
```

File: native/src/main/java/io/ballerina/lib/wso2/controlplane/ControlPlaneConstants.java [hunk 10]
```diff
@@ -7,7 +7,40 @@ public class ControlPlaneConstants {
     public static final String BAL_HOME = "balHome";
```
Comment (warunalakshitha) [inline native/src/main/java/io/ballerina/lib/wso2/controlplane/ControlPlaneConstants.java]: Shall we rename this class to Constants.java.

Comment (warunalakshitha) [reply native/src/main/java/io/ballerina/lib/wso2/controlplane/ControlPlaneConstants.java]: Shall we rename other classes to Artifacts.java, Services.Java and Listeners.Java. Current names are too length IMO

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**